support-frontend/assets/pages/paper-subscription-checkout/components/postcodeFinder.jsx

```diff
@@ -0,0 +1,84 @@
+// @flow
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { compose } from 'redux';
+
+import Button from 'components/button/button';
+import { Input } from 'components/forms/standardFields/input';
+import { Select } from 'components/forms/standardFields/select';
+import { asControlled } from 'components/forms/formHOCs/asControlled';
+import { withLabel } from 'components/forms/formHOCs/withLabel';
+import { withError } from 'components/forms/formHOCs/withError';
+
+import { type PostcodeFinderState, type Address, type PostcodeFinderActionCreators, postcodeFinderActionCreators } from './postcodeFinderReducer';
+
+import './postcodeFinder.scss';
+
+type PropTypes = {|
+  ...PostcodeFinderState,
+  ...PostcodeFinderActionCreators,
+  onPostcodeUpdate: (string) => void,
+  onAddressUpdate: (Address) => void,
+  id: string
+|};
+
+const InputWithButton = ({ onClick, ...props }) => (
+  <div className="checkout-postcode-finder">
+    <Input {...props} name="postcode" />
+    <Button type="button" appearance="greyHollow" icon={null} onClick={onClick} aria-label={null}>Find it</Button>
+  </div>
+);
+
+const ComposedInputWithButton = compose(withLabel, asControlled, withError)(InputWithButton);
+const ComposedSelect = compose(withLabel)(Select);
+
+class PostcodeFinder extends Component<PropTypes> {
+  componentDidUpdate(prevProps) {
+    if (this.selectRef && this.props.results.join() !== prevProps.results.join()) {
+      this.selectRef.focus();
+    }
+  }
+  selectRef: (Element & {focus: Function});
+  render() {
+    const {
+      id, postcode, results, setPostcode, fetchResults, error, onPostcodeUpdate, onAddressUpdate,
+    } = this.props;
+    return (
+      <div>
+        <ComposedInputWithButton
+          error={error}
+          label="Postcode"
+          onClick={fetchResults}
+          id={id}
+          setValue={(val) => {
+            setPostcode(val);
+            onPostcodeUpdate(val);
+          }}
+          value={postcode}
+        />
+        {(results.length > 0) &&
+          <ComposedSelect
+            onChange={(ev) => { if (results[ev.target.value]) { onAddressUpdate(results[ev.target.value]); } }}
```

Comment:
Maybe extract this into a standalone function for readability?

Reply:
I'd rather keep it inline because i personally find it super confusing to have event handlers extracted out of context but i've dropped a couple lines breaks!